### PR TITLE
Move parent/doc/collection back

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
As part of go/firestore-next-amendment-amendment, we decided to make `parent` once again a property of CollectionReference and DocumentReference and also to re-add `doc()` and `collection()` method to these classes (with the old methods staying as is).

Tree-shaking will be improved with https://github.com/firebase/firebase-js-sdk/pull/3678

### CODE REVIEW

File: packages/firestore/lite/src/api/reference.ts [hunk 1]
```diff
@@ -382,6 +402,31 @@ export class CollectionReference<T = firestore.DocumentData> extends Query<T>
...
   }
 
+  get parent(): DocumentReference<firestore.DocumentData> | null {
+    const parentPath = this._path.popLast();
+    if (parentPath.isEmpty()) {
+      return null;
+    } else {
+      return new DocumentReference(
+        this.firestore,
+        /* converter= */ null,
```
Comment (thebrianchen) [inline packages/firestore/lite/src/api/reference.ts:412]: nit: Can you standardize the constructor argument order for `DocumentReference` and `CollectionReference`?

Comment (thebrianchen) [reply packages/firestore/lite/src/api/reference.ts:412]: I missed this in an earlier review when you made the converter nullable, but the `converter` in the `CollectionReference` constructor can be `readonly`.

File: packages/firestore/lite/src/api/reference.ts [hunk 2]
```diff
@@ -753,3 +768,16 @@ export function newUserDataReader(firestore: Firestore): UserDataReader {
     serializer
   );
 }
+
+function validateNonEmptyArgument(
+  functionName: string,
+  argumentName: string,
+  argument?: string
+): asserts argument is string {
```
Comment (thebrianchen) [inline packages/firestore/lite/src/api/reference.ts:776]: ooooooh this is fancy :)

File: packages/firestore/lite/src/api/reference.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/lite/src/api/reference.ts:413]: Done (changed CollectionReference to match). As for `readonly`, that is done in the parent class Query and we rely on the Query's converter property here.

**Pull request merged**